src/myrtlespeech/model/hard_lstm.py

```diff
@@ -0,0 +1,419 @@
+
+        hn_out = torch.stack(hn_out)
+        cn_out = torch.stack(cn_out)
+        req_size_out = (-1, hn_out.size(2), hn_out.size(3))
+        out_states = hn_out.view(req_size_out), cn_out.view(req_size_out)
+
+        if self.batch_first:
+            output = output.transpose(0, 1)
+
+        return output, out_states
+
+
+class HardLSTMLayer(torch.nn.Module):
+    """A single-directional LSTM layer.
+
+    Args:
+        input_size: See :py:class:`HardLSTM`.
+
+        hidden_size: See :py:class:`HardLSTM`.
+
+        forget_gate_bias: See :py:class:`HardLSTM`.
+
+        reverse: A boolean. If True, this is is a reverse LSTM layer (used in
+            a bidirectional LSTM).
+    """
+
+    def __init__(
+        self,
+        input_size: int,
+        hidden_size: int,
+        forget_gate_bias: Optional[float] = None,
+        reverse: bool = False,
+    ):
+        super().__init__()
+        self.input_size = input_size
+        self.hidden_size = hidden_size
+        self.forget_gate_bias = forget_gate_bias
+        self.reverse = reverse
+        self.cell = HardLSTMCell(input_size, hidden_size)
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        stdv = 1.0 / math.sqrt(self.hidden_size)
+        for weight in self.parameters():
+            weight.data.uniform_(-stdv, stdv)
+
+        if self.forget_gate_bias is not None:
+            self.cell.bias_ih.data[
+                self.hidden_size : 2 * self.hidden_size
+            ] = self.forget_gate_bias
+            self.cell.bias_hh.data[
+                self.hidden_size : 2 * self.hidden_size
+            ] = 0.0
+
+    def forward(
+        self, x: Tensor, hx: Tuple[Tensor, Tensor]
+    ) -> Tuple[Tensor, Tuple[Tensor, Tensor]]:
+        """Performs forward pass of single LSTM layer."""
+        hx = hx[0][0], hx[1][0]
+        xs = x.unbind(0)
+        y = torch.jit.annotate(List[Tensor], [])
+        if self.reverse:
+            for t in range(len(xs) - 1, -1, -1):
+                hy, hx = self.cell(xs[t], hx)
+                y += [hy]
+            y.reverse()
+        else:
+            for t in range(len(xs)):
+                hy, hx = self.cell(xs[t], hx)
+                y += [hy]
+        (h, c) = hx
+
+        return torch.stack(y), (h.unsqueeze(0), c.unsqueeze(0))
+
+
+class HardLSTMBidirLayer(torch.nn.Module):
+    """A Bidirectional LSTM layer.
+
+    Args:
+        See :py:class:`HardLSTM`.
+    """
+
+    def __init__(
+        self,
+        input_size: int,
+        hidden_size: int,
+        forget_gate_bias: Optional[float] = None,
+    ):
+        super().__init__()
+        self.input_size = input_size
+        self.hidden_size = hidden_size
+        self.forget_gate_bias = forget_gate_bias
+
+        args = (
+            input_size,
+            hidden_size,
+            forget_gate_bias,
+        )
+        self.directions = torch.nn.ModuleList(
+            [
+                HardLSTMLayer(*args, reverse=False),
```

Comment:
See `HardLSTMLayer` comment about `reverse` - these can become `self.forward = HardLSTMLayer(...); self.backward = HardLSTMLayer(...)` and `forward` updated to cope with fiddling with input format.

Reply:
Ah yes good point. I've refactored to reflect this change

Reply:
Good point  - I've refactored to reflect this.